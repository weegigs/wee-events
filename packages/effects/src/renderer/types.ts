import * as Effect from "effect/Effect";
import { Entity, Payload } from "@weegigs/events-core";

export interface RenderFailure {
  readonly description: string;
}

export type Renderer<R, P extends Payload, A extends Payload> = Effect.Effect<(entity: Entity<P>) => Effect.Effect<A, RenderFailure>, never, R>;
export namespace Renderer {
  export type IO<P extends Payload, A extends Payload = Payload> = Renderer<unknown, P, A>;
}

export type AnyRenderer<R> = Renderer<R, Payload, Payload>;
export namespace AnyRenderer {
  export type IO = AnyRenderer<unknown>;

  export const from = <R, P extends Payload, A extends Payload>(source: Renderer<R, P, A>): AnyRenderer<R> => {
    return source as unknown as AnyRenderer<R>;
  };
}
