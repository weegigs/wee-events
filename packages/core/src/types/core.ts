import * as z from "zod";

export type Primitive = string | number | boolean;
export type Value = Primitive | Payload | List;
export type EmptyPayload = { [member: string]: never };
export type Payload = { [member: string]: Value } | EmptyPayload;
export type List = Primitive[] | Payload[] | List[];

export namespace Payload {
  const primitive = z.union([z.string(), z.number(), z.boolean()]);
  const list: z.Schema<List> = z.lazy(() => z.union([z.array(primitive), z.array(payload), z.array(list)]));
  const value: z.Schema<Value> = z.lazy(() => z.union([primitive, list, payload]));
  const payload: z.Schema<Payload> = z.lazy(() => z.record(value));

  export const schema = payload;
  export const parse = (value: unknown): Payload => schema.parse(value);
}

export namespace EmptyPayload {
  export const schema = z.record(z.never());
}

export type AggregateId<Type extends string = string> = {
  type: Type;
  key: string;
};

export namespace AggregateId {
  export const schema = <Type extends string>(type: Type): z.Schema<AggregateId<Type>> =>
    z.object({
      type: z.literal(type),
      key: z.string().min(1),
    }) as unknown as z.Schema<AggregateId<Type>>;

  export const encode = (aggregate: AggregateId): string => `${aggregate.type}.${aggregate.key}`;
  export const decoder =
    (type: string) =>
    (aggregate: string): AggregateId => {
      const [_type, ...key] = aggregate.split(".");
      return schema(type).parse({ type: _type, key: key.join(".") });
    };

  export const parser =
    <Type extends string>(type: Type) =>
    (value: unknown) =>
      schema(type).parse(value);

  export function create<const T extends string>(type: T, key: string): AggregateId<T> {
    return { type, key } as const;
  }
}

export interface DomainEvent<Type extends string = string, Data extends Payload = Payload> {
  type: Type;
  data: Data;
}

export namespace DomainEvent {
  export const schema = <S extends z.Schema>(type: string, data: S) =>
    z.object({
      type: z.literal(type),
      data,
    });
}

export interface RecordedEvent<E extends DomainEvent = DomainEvent> {
  /**
   * The aggregate that events belongs to.
   */
  aggregate: AggregateId;

  /**
   * The revision for the aggregate.
   */
  revision: string;

  /**
   * Unique identifier representing this event. ULID format.
   */
  id: string;

  /**
   * Type of this event.
   */
  type: E["type"];

  /**
   * Representing when this event was created in the database system.
   */
  timestamp: string;

  /**
   * Data of this event.
   */
  data: E["data"];

  /**
   * Representing the metadata associated with this event.
   */
  metadata: RecordedEvent.Metadata;
}

export namespace RecordedEvent {
  export type Metadata = {
    causationId?: string;
    correlationId?: string;
  };
}

export type Revision = string;

export namespace Revision {
  export type Initial = "00000000000000000000000000";
  export const Initial = "00000000000000000000000000";

  export const isInitial = (revision: Revision): boolean => revision === Initial;

  export const schema = z.string().min(26).max(26);
}
