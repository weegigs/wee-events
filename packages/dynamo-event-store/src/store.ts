import _ from "lodash";
import * as z from "zod";

import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, QueryCommand, TransactWriteCommand, QueryCommandOutput } from "@aws-sdk/lib-dynamodb";

import { ChangeSet } from "./changeset";

// Simple error classification functions to replace deprecated @aws-sdk/service-error-classification
const isRetryableByTrait = (error: any) => error?.retryable === true;
const isClockSkewError = (error: any) => error?.name === "ClockSkew";
const isThrottlingError = (error: any) => error?.name === "ThrottlingException" || error?.name === "ProvisionedThroughputExceededException";
const isTransientError = (error: any) => error?.name === "ServiceUnavailable" || error?.name === "InternalServerError";

import { retry } from "@weegigs/events-common";
import { Cypher, Tokenizer, no } from "@weegigs/events-cypher";

import {
  DomainEvent,
  EventStore,
  RecordedEvent,
  Revision,
  AggregateId,
  ExpectedRevisionConflictError,
  RevisionConflictError,
  Payload,
} from "@weegigs/events-core";

export namespace DynamoEventStore {
  export interface Options {
    client: () => DynamoDBDocumentClient;
    cypher: Cypher;
    tokenizer: Tokenizer;
  }
}

const defaults: DynamoEventStore.Options = {
  client: () =>
    DynamoDBDocumentClient.from(new DynamoDBClient({}), {
      marshallOptions: {
        removeUndefinedValues: true,
      },
    }),
  cypher: no.cypher,
  tokenizer: no.tokenizer,
};

type LatestRecord = {
  pk: string;
  sk: string;
  revision: string;
  timestamp: string;
};

function latestFor(changeSet: ChangeSet): LatestRecord {
  return {
    pk: changeSet.pk,
    sk: `latest-revision`,
    revision: changeSet.revision,
    timestamp: changeSet.timestamp,
  };
}

function latestCondition(changeset: ChangeSet, expectedVersion?: string) {
  if (undefined === expectedVersion) {
    return {
      ConditionExpression: "attribute_not_exists(#revision) Or #revision < :revision",
      ExpressionAttributeNames: { "#revision": "revision" },
      ExpressionAttributeValues: { ":revision": changeset.revision },
    };
  }

  if (expectedVersion === Revision.Initial) {
    return {
      ConditionExpression: "attribute_not_exists(#revision)",
      ExpressionAttributeNames: { "#revision": "revision" },
    };
  }

  return {
    ConditionExpression: "#revision = :revision",
    ExpressionAttributeNames: { "#revision": "revision" },
    ExpressionAttributeValues: { ":revision": expectedVersion },
  };
}

const isRetryableError = (error: unknown) => {
  return isRetryableByTrait(error as any) || isClockSkewError(error as any) || isThrottlingError(error as any) || isTransientError(error as any);
};

const isConditionCheckFailed = (error: unknown) =>
  (error as any)?.name === "TransactionCanceledException" &&
  (error as any).CancellationReasons?.[0]?.Code === "ConditionalCheckFailed";

export class DynamoEventStore implements EventStore {
  private readonly $client: DynamoDBDocumentClient;
  private readonly $table: string;
  private readonly $encoder: ChangeSet.EventEncoder;
  private readonly $decoder: ChangeSet.ChangeSetDecoder;

  constructor(table: string, options: Partial<DynamoEventStore.Options> = {}) {
    const { client, cypher, tokenizer } = { ...defaults, ...options };

    this.$client = client();
    this.$table = table;
    this.$encoder = ChangeSet.encoder(cypher, tokenizer);
    this.$decoder = ChangeSet.decoder(cypher);
  }

  private $read = async (
    key: string,
    decrypt: boolean,
    _after?: Revision,
    cursor: any = undefined
  ): Promise<{ events: RecordedEvent[]; next: any }> => {
    const query = new QueryCommand({
      TableName: this.$table,
      ScanIndexForward: true,
      ExclusiveStartKey: cursor,
      KeyConditionExpression: "#pk = :aggregate And begins_with(#sk, :change_set_prefix)",
      ExpressionAttributeNames: {
        "#pk": "pk",
        "#sk": "sk",
      },
      ExpressionAttributeValues: {
        ":aggregate": key,
        ":change_set_prefix": "change-set#",
      },
    });

    const { Items, LastEvaluatedKey } = await retry(
      async () => this.$client.send(query),
      (e: Error | QueryCommandOutput) => {
        return _.isError(e) ? isRetryableError(e) : false;
      },
      { limit: 5, delay: 37 }
    );

    const events: RecordedEvent[][] = await Promise.all(
      (Items ?? [])
        .map((item: any) => ChangeSet.schema.parse(item))
        .map((changeSet) => this.$decoder(changeSet, decrypt))
    );

    return { events: events.flat(), next: LastEvaluatedKey };
  };

  load = async (aggregate: AggregateId, options: EventStore.LoadOptions = {}): Promise<RecordedEvent[]> => {
    const key = AggregateId.encode(aggregate);
    const { decrypt, afterRevision: after } = { decrypt: false, ...options };

    let events: RecordedEvent[] = [];
    let cursor: any | undefined = undefined;
    do {
      const { events: page, next } = await retry(
        () => this.$read(key, decrypt, after, cursor),
        (result) => _.isError(result) && isRetryableError(result)
      );

      events = events.concat(page);
      cursor = next;
    } while (cursor);

    return events;
  };

  publish = async (
    aggregate: AggregateId,
    events: DomainEvent | DomainEvent[],
    options: EventStore.PublishOptions = {}
  ): Promise<Revision> => {
    const _events = Array.isArray(events) ? events : [events];
    if (_events.length === 0) {
      throw new Error(`expected at least one event to be published`);
    }

    const { expectedRevision } = options;

    try {
      return await retry(
        async (): Promise<Revision> => this.$publish(aggregate, _events, options),
        (e: Error | Revision) => {
          if (!_.isError(e)) {
            return false;
          }
          if (
            e.name === "TransactionCanceledException" &&
            (e as any).CancellationReasons?.[0]?.Code === "ConditionalCheckFailed"
          ) {
            return expectedRevision === undefined;
          }

          return isRetryableError(e);
        },
        { limit: 5, delay: 37 }
      );
    } catch (e: unknown) {
      if (_.isError(e) && isConditionCheckFailed(e)) {
        if (expectedRevision !== undefined) {
          throw new ExpectedRevisionConflictError(expectedRevision);
        }

        throw new RevisionConflictError();
      }

      throw e;
    }
  };

  private async $publish(
    aggregate: AggregateId,
    events: DomainEvent[],
    options: EventStore.PublishOptions = {}
  ): Promise<Revision> {
    const { encrypt, expectedRevision } = options;

    const recorded: RecordedEvent[] = await Promise.all(
      events
        .map((event) => {
          // Validate that the event has the proper DomainEvent structure
          const domainEventSchema = z.object({
            type: z.string().min(1),
            data: Payload.schema
          });
          return domainEventSchema.parse(event);
        })
        .map((event) =>
          this.$encoder(aggregate, event, _.pick(options, "correlationId", "causationId"), encrypt ?? false)
        )
    );

    const changeSet = await ChangeSet.create(aggregate, recorded);
    return this.$write(changeSet, expectedRevision);
  }

  private async $write(changeSet: ChangeSet, expectedRevision?: string): Promise<Revision> {
    const TableName = this.$table;

    const latest = latestFor(changeSet);
    const latestCheck = latestCondition(changeSet, expectedRevision);

    const write = new TransactWriteCommand({
      TransactItems: [
        {
          Put: {
            TableName,
            Item: latest,
            ...latestCheck,
          },
        },
        {
          Put: {
            TableName,
            Item: changeSet,
          },
        },
      ],
    });

    await this.$client.send(write);

    return changeSet.revision;
  }
}
