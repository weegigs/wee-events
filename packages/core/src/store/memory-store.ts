import { DateTime } from "luxon";
import { monotonicFactory } from "ulid";

import {
  AggregateId,
  DomainEvent,
  RecordedEvent,
  Revision,
} from "../types";
import { EventStore } from "./types";

const ulid = monotonicFactory();

export class MemoryStore implements EventStore {
  aggregates: Record<string, RecordedEvent[]> = {};

  clear() {
    this.aggregates = {};
  }

  async load(aggregate: AggregateId): Promise<RecordedEvent[]> {
    return this.aggregates[AggregateId.encode(aggregate)] ?? [];
  }

  async publish(aggregate: AggregateId, events: DomainEvent | DomainEvent[]): Promise<Revision> {
    const current = this.aggregates[AggregateId.encode(aggregate)] ?? [];
    const revision = current[current.length - 1]?.revision ?? Revision.Initial;

    const e = Array.isArray(events) ? events : [events];
    if (e.length == 0) {
      return revision;
    }

    const recorded: RecordedEvent[] = e.map((event) => {
      const revision = ulid()
      const timestamp = DateTime.now().toISO()

      if (timestamp == null) {
        throw new Error("unexpected invalid timestamp")
      }

      return {
        ...event,
        id: revision,
        revision,
        aggregate: aggregate,
        timestamp,
        metadata: {},
      };
    });

    const updated = current.concat(...recorded);
    this.aggregates[AggregateId.encode(aggregate)] = updated;

    return updated[updated.length - 1].revision;
  }
}
