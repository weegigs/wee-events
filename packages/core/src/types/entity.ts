import * as z from "zod";

import { AggregateId, Payload, Revision } from "./core";

export interface Entity<T extends Payload = Payload> {
  aggregate: AggregateId;
  type: string;
  revision: Revision;
  state: T;
}

export namespace Entity {
  export const schema = <Type extends string, Data extends z.ZodTypeAny>(type: Type, state: Data) =>
    z.object({
      aggregate: AggregateId.schema(type),
      type: z.string().min(1),
      revision: Revision.schema,
      state,
    });

  export function create<State extends Payload>(
    aggregate: AggregateId,
    type: string,
    revision: Revision,
    state: State
  ): Entity<State> {
    return { aggregate, type, revision, state } as const;
  }
}
