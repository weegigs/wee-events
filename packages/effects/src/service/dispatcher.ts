import * as wee from "@weegigs/events-core";

import * as z from "zod";

import * as T from "@effect-ts/core/Effect";
import { pipe } from "@effect-ts/core/Function";

type Payload = wee.Payload;

export type Command = wee.Payload;
export type State = Payload;

export type Handler<R, E, S extends Payload, C extends Command> = (
  entity: wee.Entity<S>,
  command: C
) => T.Effect<R, E, void>;

export type ValidationError = z.ZodError;

export class CommandValidationError extends Error {
  constructor(public readonly error: ValidationError) {
    super("command validation failed");
    this.name = "CommandValidationError";

    Object.setPrototypeOf(this, CommandValidationError.prototype);
  }
}

export class HandlerNotFound extends Error {
  constructor(public readonly path: string) {
    super(`Handler not found for path ${path}`);
    this.name = "HandlerNotFound";

    Object.setPrototypeOf(this, HandlerNotFound.prototype);
  }
}

const validate = <T>(schema: z.Schema<T>, value: T): T.IO<CommandValidationError, T> => {
  return T.suspend(() => {
    const result = schema.safeParse(value);
    if (!result.success) {
      return T.fail(new CommandValidationError(result.error));
    }

    return T.succeed(result.data);
  });
};

function validated<R, E, S extends Payload, C extends Payload>(
  schema: z.Schema<C>,
  handler: Handler<R, E, S, C>
): Handler<R, E | CommandValidationError, S, C> {
  return (entity: wee.Entity<S>, command: C) =>
    pipe(
      T.do,
      T.bind("command", () => validate(schema, command)),
      T.bind("result", () => handler(entity, command)),
      T.map(({ result }) => result)
    );
}

export type Target<T extends Payload = Payload> = wee.Entity<T> | wee.AggregateId;

export interface Dispatcher<R, E, State extends Payload> {
  dispatch(
    path: string,
    entity: Target<State>,
    command: Command
  ): T.Effect<R, E | HandlerNotFound | CommandValidationError, void>;

  handler: <R1, E1, C extends Command>(
    path: string,
    schema: z.Schema<C>,
    h: Handler<R1, E1, State, C>
  ) => Dispatcher<R & R1, E | E1, State>;

  commands: () => Record<string, z.Schema<Payload>>;
}

type Handlers<R, E, State extends Payload> = Record<
  string,
  { schema: z.Schema<unknown>; handler: Handler<R, E, State, Command> }
>;

export namespace Dispatcher {
  const $make = <R, E, State extends Payload>(handlers: Handlers<R, E, State> = {}): Dispatcher<R, E, State> => {
    const dispatch = (
      path: string,
      entity: wee.Entity<State>,
      command: Command
    ): T.Effect<R, E | HandlerNotFound | CommandValidationError, void> => {
      const handler = handlers[path];
      if (undefined == handler) {
        return T.fail(new HandlerNotFound(path));
      }

      return handler.handler(entity, command);
    };

    const handler = <R1, E1, C extends Command>(
      path: string,
      schema: z.Schema<C>,
      h: Handler<R1, E1, State, C>
    ): Dispatcher<R & R1, E | E1, State> => {
      const y = {
        [path]: { schema, handler: validated(schema, h) },
      };

      const x = {
        ...handlers,
        ...y,
      } as Handlers<R & R1, E | E1, State>;

      return $make(x);
    };

    const commands = () => {
      const result: Record<string, z.Schema> = {};

      for (const path in handlers) {
        result[path] = handlers[path].schema;
      }

      return result;
    };

    return {
      dispatch,
      commands,
      handler,
    };
  };

  export const create = <State extends Payload>(): Dispatcher<unknown, never, State> => $make({});
}
