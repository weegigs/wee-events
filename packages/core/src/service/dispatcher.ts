import * as z from "zod";

import { EventStore } from "../store";
import { Entity, Payload } from "../types";
import { State } from "./state";

export type Command = Payload;

export type Environment = Record<string, unknown>;

export type Publisher = { publish: EventStore.Publisher };

export type Handler<R extends Environment, S extends State, C extends Command> = (
  environment: R,
  entity: Entity<S>,
  command: C
) => Promise<void>;

export type ValidationError = z.ZodError;

export class CommandValidationError extends Error {
  constructor(public readonly error: ValidationError) {
    super("command validation failed");
    this.name = "CommandValidationError";

    Object.setPrototypeOf(this, CommandValidationError.prototype);
  }
}

export class HandlerNotFound extends Error {
  constructor(public readonly path: string) {
    super(`Handler not found for path ${path}`);
    this.name = "HandlerNotFound";

    Object.setPrototypeOf(this, HandlerNotFound.prototype);
  }
}

const validate = <T>(schema: z.Schema<T>, value: unknown): T => {
  const result = schema.safeParse(value);

  if (!result.success) {
    throw new CommandValidationError(result.error);
  }

  return result.data;
};

export interface Dispatcher<S extends State> {
  dispatch(path: string, entity: Entity<S>, command: Command): Promise<void>;
}

export namespace Dispatcher {
  export interface Description {
    commands: Record<string, z.Schema<Payload>>;
  }
}

export interface DispatcherDescription<R extends Environment, S extends State> {
  commands(): Record<string, z.Schema<Payload>>;
  dispatcher(environment: R): Dispatcher<S>;
}

export namespace DispatcherDescription {
  class Factory<R extends Environment, S extends State> {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    private handlers: Record<string, { schema: z.Schema<any>; handler: Handler<any, S, any> }> = {};

    constructor(
      path: string,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      schema: z.Schema<any>,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      handler: Handler<R, S, any>,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      existing: Record<string, { schema: z.Schema<any>; handler: Handler<any, S, any> }> = {}
    ) {
      this.handlers = {
        ...existing,
        [path]: { schema, handler },
      };
    }

    handler: <R1 extends Environment, C extends Payload>(
      path: string,
      schema: z.ZodType<C, z.ZodTypeDef, C>,
      handler: Handler<R1, S, C>
    ) => Factory<R & R1, S> = <R1 extends Environment, C extends Payload>(
      path: string,
      schema: z.ZodType<C, z.ZodTypeDef, C>,
      handler: Handler<R1, S, C>
    ) => {
      return new Factory<R & R1, S>(path, schema, handler, this.handlers);
    };

    description: () => DispatcherDescription<R, S> = () => {
      return {
        dispatcher: (environment: R) => ({
          dispatch: async (path: string, entity: Entity<S>, command: Payload) => {
            const handler = this.handlers[path];

            if (!handler) {
              throw new HandlerNotFound(path);
            }

            await handler.handler(environment, entity, validate(handler.schema, command));
          },
        }),
        commands: () => Object.fromEntries(Object.entries(this.handlers).map(([path, { schema }]) => [path, schema])),
      };
    };
  }

  /**
   * Creates a dispatch factory that can be used to create a dispatcher
   *
   * @param path the path to connect the handler to
   * @param schema the zod schema used to validate the command
   * @param handler the handler used to execute the command
   * @returns a dispatch factory that can be used to create a dispatcher
   */
  export function handler<R extends Environment, S extends State, P extends string, C extends Command>(
    path: P,
    schema: z.Schema<C>,
    handler: Handler<R, S, C>
  ): Factory<R, S> {
    return new Factory(path, schema, handler, {});
  }
}

type HF<R extends Environment, P extends string, S extends State, C extends Command> = {
  path: P;
  schema: z.ZodType<C, z.ZodTypeDef, C>;
  handler: Handler<R, S, C>;
};

export function cf<R extends Environment, P extends string, S extends State, C extends Command>(
  hf: HF<R, P, S, C>
): (environment: R, path: P, state: Entity<S>, command: C) => Promise<void> {
  return async (environment: R, _path: P, state: Entity<S>, command: C) => {
    hf.handler(environment, state, validate(hf.schema, command));
  };
}
