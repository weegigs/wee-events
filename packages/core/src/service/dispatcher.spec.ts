import { describe, it, expect, beforeEach } from "vitest";
import { z } from "zod";

import { EventStore } from "../store";
import { AggregateId, DomainEvent, Entity, Revision } from "../types";
import { CommandValidationError, DispatcherDescription, HandlerNotFound, Publisher } from "./dispatcher";

const withLogger = <R extends Publisher>(environment: R): R & { logger: (message: string) => void } => {
  const logger = (message: string) => {
    console.log(message);
  };

  return Object.assign({}, environment, { logger });
};

type Logger = { logger: (message: string) => void };

type Count = {
  current: number;
};

type Add = {
  amount: number;
};

namespace Add {
  export const schema = z.object({
    amount: z.number(),
  });
}

const add = async (environment: Publisher, entity: Entity<Count>, command: Add) => {
  const { publish } = environment;

  if (command.amount != 0) {
    await publish(entity.aggregate, { type: "added", data: { amount: command.amount } });
  }
};

type Subtract = {
  amount: number;
};

namespace Subtract {
  export const schema = Add.schema;
}

const subtract = async (environment: Publisher, entity: Entity<Count>, command: Subtract) => {
  const { publish } = environment;

  if (command.amount != 0) {
    await publish(entity.aggregate, { type: "subtracted", data: { amount: -command.amount } });
  }
};

type Clear = Record<string, never>;

namespace Clear {
  export const schema = z.object({});
}

const clear = async (environment: Logger & Publisher, entity: Entity<Count>, _: Clear) => {
  const { publish, logger } = environment;

  if (entity.state.current != 0) {
    logger(`Clearing ${entity.aggregate.key}`);
    await publish(entity.aggregate, { type: "cleared", data: {} });
  }
};

describe("dispatcher", () => {
  let events: Record<string, DomainEvent[]> = {};

  beforeEach(() => {
    events = {};
  });

  const publish: EventStore.Publisher = async (aggregate: AggregateId, event: DomainEvent) => {
    const id = AggregateId.encode(aggregate);
    const _events = events[id] || [];
    events[id] = [..._events, event];

    return "";
  };

  const environment = withLogger({ publish });

  const target = Entity.create(AggregateId.create("counter", "1"), "counter", Revision.Initial, { current: 0 });

  it("should allow a dispatcher to be created with a single handler", async () => {
    const dispatcher = DispatcherDescription.handler("add", Add.schema, add).description().dispatcher(environment);

    await dispatcher.dispatch("add", target, { amount: 1 });

    expect(events[AggregateId.encode(target.aggregate)]).toHaveLength(1);
    expect(events).toMatchSnapshot();
  });

  it("should allow a dispatcher to be created with multiple handlers", async () => {
    const dispatcher = DispatcherDescription.handler("add", Add.schema, add)
      .handler("subtract", Subtract.schema, subtract)
      .handler("clear", Clear.schema, clear)
      .description()
      .dispatcher(environment);

    await dispatcher.dispatch("add", target, { amount: 1 });
    await dispatcher.dispatch("subtract", target, { amount: 2 });

    expect(events[AggregateId.encode(target.aggregate)]).toHaveLength(2);
    expect(events).toMatchSnapshot();
  });

  it("should fail with a handler not found error if a command is sent to an unknown path", async () => {
    const dispatcher = DispatcherDescription.handler("add", Add.schema, add).description().dispatcher(environment);

    await expect(dispatcher.dispatch("multiply", target, { amount: 1 })).rejects.toThrowError(HandlerNotFound);
  });

  it("should fail with a validation error if a command payload is incorrect", async () => {
    const dispatcher = DispatcherDescription.handler("add", Add.schema, add).description().dispatcher(environment);

    await expect(dispatcher.dispatch("add", target, { quantity: 1 })).rejects.toThrowError(CommandValidationError);
  });

  it("should allow describe the paths and commands available", async () => {
    const description = DispatcherDescription.handler("add", Add.schema, add)
      .handler("subtract", Subtract.schema, subtract)
      .handler("clear", Clear.schema, clear)
      .description();

    expect(Object.keys(description.commands())).toHaveLength(3);
    expect(Object.keys(description.commands()).sort()).toEqual("add,clear,subtract".split(","));
  });
});