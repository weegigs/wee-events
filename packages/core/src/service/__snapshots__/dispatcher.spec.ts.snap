// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dispatcher > should allow a dispatcher to be created with a single handler 1`] = `
{
  "counter.1": [
    {
      "data": {
        "amount": 1,
      },
      "type": "added",
    },
  ],
}
`;

exports[`dispatcher > should allow a dispatcher to be created with multiple handlers 1`] = `
{
  "counter.1": [
    {
      "data": {
        "amount": 1,
      },
      "type": "added",
    },
    {
      "data": {
        "amount": -2,
      },
      "type": "subtracted",
    },
  ],
}
`;
